test/units/executor/test_playbook_executor.py

```diff
@@ -93,54 +90,109 @@ def test_get_serialized_batches(self):
+            ['host4', 'host5'],
+            ['host6', 'host7'],
+            ['host8', 'host9']
+        ]
 
         playbook = Playbook.load(pbe._playbooks[3], variable_manager=mock_var_manager, loader=fake_loader)
         play = playbook.get_plays()[0]
         play.post_validate(templar)
         mock_inventory.get_hosts.return_value = ['host0', 'host1', 'host2', 'host3', 'host4', 'host5', 'host6', 'host7', 'host8', 'host9']
-        self.assertEqual(
-            pbe._get_serialized_batches(play),
-            [['host0'], ['host1', 'host2'], ['host3', 'host4', 'host5'], ['host6', 'host7', 'host8'], ['host9']]
-        )
+        assert pbe._get_serialized_batches(play) == [
+            ['host0'],
+            ['host1', 'host2'],
+            ['host3', 'host4', 'host5'],
+            ['host6', 'host7', 'host8'],
+            ['host9']
+        ]
 
         playbook = Playbook.load(pbe._playbooks[4], variable_manager=mock_var_manager, loader=fake_loader)
         play = playbook.get_plays()[0]
         play.post_validate(templar)
         mock_inventory.get_hosts.return_value = ['host0', 'host1', 'host2', 'host3', 'host4', 'host5', 'host6', 'host7', 'host8', 'host9']
-        self.assertEqual(pbe._get_serialized_batches(play), [['host0'], ['host1', 'host2'], ['host3', 'host4', 'host5', 'host6', 'host7', 'host8', 'host9']])
+        assert pbe._get_serialized_batches(play) == [
+            ['host0'],
+            ['host1', 'host2'],
+            ['host3', 'host4', 'host5', 'host6', 'host7', 'host8', 'host9']
+        ]
 
         # Test when serial percent is under 1.0
         playbook = Playbook.load(pbe._playbooks[2], variable_manager=mock_var_manager, loader=fake_loader)
         play = playbook.get_plays()[0]
         play.post_validate(templar)
         mock_inventory.get_hosts.return_value = ['host0', 'host1', 'host2']
-        self.assertEqual(pbe._get_serialized_batches(play), [['host0'], ['host1'], ['host2']])
+        assert pbe._get_serialized_batches(play) == [['host0'], ['host1'], ['host2']]
 
         # Test when there is a remainder for serial as a percent
         playbook = Playbook.load(pbe._playbooks[2], variable_manager=mock_var_manager, loader=fake_loader)
         play = playbook.get_plays()[0]
         play.post_validate(templar)
         mock_inventory.get_hosts.return_value = ['host0', 'host1', 'host2', 'host3', 'host4', 'host5', 'host6', 'host7', 'host8', 'host9', 'host10']
-        self.assertEqual(
-            pbe._get_serialized_batches(play),
-            [['host0', 'host1'], ['host2', 'host3'], ['host4', 'host5'], ['host6', 'host7'], ['host8', 'host9'], ['host10']]
+        assert pbe._get_serialized_batches(play) == [
+            ['host0', 'host1'],
+            ['host2', 'host3'],
+            ['host4', 'host5'],
+            ['host6', 'host7'],
+            ['host8', 'host9'],
+            ['host10']
+        ]
+
+    def test_generate_retry_inventory_success(self, mocker, tmp_path):
```

Comment:
It would be better to write this test, and the following, as integration tests. That would avoid the use of mocking and internal function calls, making the test useful even after refactoring.

Use the `ssh` connection to localhost with a bad port and set `ANSIBLE_RETRY_FILES_ENABLED=1` .
To get the write failure, setting `ANSIBLE_RETRY_FILES_SAVE_PATH=/dev/null` would work.